packages/enhanced/src/lib/container/HoistContainerReferencesPlugin.ts

```diff
@@ -4,27 +4,123 @@ import type {
  * @constructor
  */
-export class HoistContainerReferences implements WebpackPluginInstance {
+export class HoistContainerReferencesPlugin implements WebpackPluginInstance {
+  integrateChunks(
+    chunkA: Chunk,
+    chunkB: Chunk,
+    compilation: Compilation,
+  ): void {
+    const { chunkGraph, compiler } = compilation;
+    // do not sort chunk by smallest name, this will cause non-deterministic chunk integration
+
+    // if (chunkA.name && chunkB.name) {
+    //   if (
+    //     chunkGraph.getNumberOfEntryModules(chunkA) > 0 ===
+    //     chunkGraph.getNumberOfEntryModules(chunkB) > 0
+    //   ) {
+    //     // When both chunks have entry modules or none have one, use
+    //     // shortest name
+    //     if (chunkA.name.length !== chunkB.name.length) {
+    //       chunkA.name =
+    //         chunkA.name.length < chunkB.name.length ? chunkA.name : chunkB.name;
+    //     } else {
+    //       chunkA.name = chunkA.name < chunkB.name ? chunkA.name : chunkB.name;
+    //     }
+    //   } else if (chunkGraph.getNumberOfEntryModules(chunkB) > 0) {
+    //     // Pick the name of the chunk with the entry module
+    //     chunkA.name = chunkB.name;
+    //   }
+    // } else if (chunkB.name) {
+    //   chunkA.name = chunkB.name;
+    // }
+    //
+    // // Merge id name hints
+    for (const hint of chunkB.idNameHints) {
+      chunkA.idNameHints.add(hint);
+    }
+
+    // Merge runtime
+    //@ts-ignore
+    chunkA.runtime = runtime.mergeRuntime(chunkA.runtime, chunkB.runtime);
+
+    // getChunkModules is used here to create a clone, because disconnectChunkAndModule modifies
+    for (const module of chunkGraph.getChunkModules(chunkB)) {
+      // if (chunkB.name !== 'federation-runtime') {
+      chunkGraph.disconnectChunkAndModule(chunkB, module);
+      // }
+      chunkGraph.connectChunkAndModule(chunkA, module);
+    }
+
+    for (const [module, chunkGroup] of Array.from(
+      chunkGraph.getChunkEntryModulesWithChunkGroupIterable(chunkB),
+    )) {
+      // chunkGraph.disconnectChunkAndEntryModule(chunkB, module);
+      chunkGraph.connectChunkAndModule(chunkA, module);
+      // chunkGraph.connectChunkAndEntryModule(chunkA, module,chunkGroup);
+    }
+
+    for (const chunkGroup of chunkB.groupsIterable) {
+      chunkGroup.replaceChunk(chunkB, chunkA);
+      chunkA.addGroup(chunkGroup);
+      chunkB.removeGroup(chunkGroup);
+    }
+    compiler.webpack.ChunkGraph.clearChunkGraphForChunk(chunkB);
+  }
+
   apply(compiler: Compiler): void {
-    compiler.hooks.thisCompilation.tap(
-      'HoistContainerReferences',
+    compiler.hooks.compilation.tap(
+      'HoistContainerReferencesPlugin',
       (compilation: Compilation) => {
-        compilation.hooks.afterOptimizeChunks.tap(
-          'HoistContainerReferences',
+        compilation.hooks.optimizeChunks.tap(
+          {
+            name: 'HoistContainerReferencesPlugin',
+            stage: 10, // advanced stage chunk optimization
```

Comment:
Integrate chunks at latest stage of chunk optimization phase, this helps to avoid tree shaking or marking used modules as unused modules due to runtime module being the one who consumes it